### DISCUSSION TITLE
Shareable iFrame Support and Budibase WordPress Plugin

### DISCUSSION THREAD

**Discussion by switchupcb** [Budibase/budibase#9792] [Ideas]
This idea aims to meet the following objectives.
1. Expand the market share of Budibase among website developers and end users.
2. Make it easier to use Budibase in external websites.

## Context
WordPress powers 42% of the internet. Alternatives to WordPress have been created, but fail for various reasons. As an example, read why [Jamstack Won't Replace WordPress](https://switchupcb.medium.com/jamstack-wont-replace-wordpress-52b10757a43). Budibase is not an alternative to WordPress, but it isn't compatible yet either. However, the work to make it compatible is very simple. Then, there is an additional opportunity that Budibase is not taking advantage of.

## Shareable iFrame Support
Websites are able to embed other applications using iFrames. However, a quick search about using iFrames with Budibase shows that the feature isn't supported. Here are the existing discussion on the subject within this repository.
- https://github.com/Budibase/budibase/discussions/3936 which references [appsmith embed into existing application](https://docs.appsmith.com/advanced-concepts/embed-appsmith-into-existing-application).
- https://github.com/Budibase/budibase/issues/5659 (Embed budibase into existing web app).
- https://github.com/Budibase/budibase/discussions/8090 (Integrate budibase app to another app).

A comment by @wooparadog indicates that iFrame embeds are currently flawed since Budibase strips cookies from the user. A comment by @oswaldB implies that it can be done with the configuration of the reverse proxy, but this is — based on Budibase's competitors — unnecessary work for the end user. Implementing shareable iFrame support could be as simple as adding the documentation to do it without issue. Doing so would allow every web developer to use Budibase in their work, which expands the market of Budibase.

_While this feature is not necessarily related to a WordPress Plugin, it may help in its implementation. Something as simple as allowing Budibase components to be used in a WordPress website would expand Budibase's market._

## Budibase Pagebuilder (WordPress Plugin)
Modern WordPress websites are being created with a **page builders**, which function in a similar manner to the Budibase application builder. WordPress page builders work by disabling the WP Theme and any other unnecessary files, then allowing the end user to create custom templates, components, and widgets. This allows the end user to create a custom website, while also maintaining the benefits of WordPress. The following software is an example of notable WordPress page builders.
1. [Oxygen](https://oxygenbuilder.com/) ([Demo](https://oxygenbuilder.com/documentation/getting-started/getting-started-tutorial/))
2. [Bricks](https://bricksbuilder.io)
3. Elementor
4. etc

**Here is why Budibase will benefit from creating a page builder of their own for the WordPress ecosystem.**
1. The users are willing to pay for the software.  Check the pricing of each page builder above for an example of how much money users are willing to spend on this feature. Budibase's current users are developers who don't even want to give out free stars.
2. The Budibase page builder is already a product of Budibase. There shouldn't be too much effort required to maintain a version that is compatible with WordPress; or simply create a plugin that makes the connection between the Budibase instance and the WordPress instance easy to set up.
3. Budibase will become directly accessible to 42% of websites that exist on the internet.

**Comment by wooparadog** [comment]
Wow, thanks for the mention. I'll update some findings after my last comment.

1. My original problem is that budibase will strip cookies in its http api data endpoint, not  iframe. 
1. It's not wrong for app to strip cookies in subsequent request to another server, and for security reasons, budibase is doing it right by using this behavior by default. 
1. Last time I checked, there's a walkaround for my use case(using document.cookie or something).  But it's a pain in the ass. 
1. I STRONGLY agree budibase should improve its abilities integrating with both internal and external tools 

Thanks again for crediting my short comment.

**Comment by autonia** [comment]
This. Was very surprised there isn't an option to use an iframe to embed an app. I'm vetting various "internal tool" builders for our org and am walking away from Budibase based on this factor alone.

**Comment by dosandco** [comment]
I notice Drona HQ released this feature yesterday on ProductHunt: https://www.dronahq.com/embed-apps/ - would definitely be a cool BB feature!

**Comment by ConorWebb96** [comment]
Hey everyone,

This is in the pipeline to be added to Budibase, currently you are able to do this with selfhosted by configuring your nginx/htaccess settings. (Examples of how to do this i've provided below.) We are hoping to extend this to cloud services as well over the coming iterations. (No exact timeline for this, however this is a feature thats heavily requested and will be added to Budibase.)

.**htaccess**
```
Header set X-Frame-Options "ALLOW-FROM https://yourdomain.com/"
```

**nginx**
```
add_header X-Frame-Options "ALLOW-FROM https://yourdomain.com/";
```

I hope this information helps!
